engine-types/src/types.rs

```diff
@@ -104,6 +108,114 @@ impl Mul<EthGas> for u64 {
     }
 }
 
+#[derive(
+    Default, Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd, BorshSerialize, BorshDeserialize,
+)]
+/// Engine `fee` type which wraps an underlying u128.
+pub struct Fee(u128);
+
+impl Display for Fee {
+    fn fmt(&self, f: &mut Formatter<'_>) -> crate::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl Fee {
+    /// Constructs a new `Fee` with a given u128 value.
+    pub const fn new(fee: u128) -> Fee {
+        Self(fee)
+    }
+
+    /// Consumes `Fee` and returns the underlying type.
+    pub fn into_u128(self) -> u128 {
+        self.0
+    }
+}
+
+impl Add<Fee> for Fee {
+    type Output = Fee;
+
+    fn add(self, rhs: Fee) -> Self::Output {
+        Fee(self.0 + rhs.0)
+    }
+}
+
+impl Add<Fee> for u128 {
+    type Output = Fee;
+
+    fn add(self, rhs: Fee) -> Self::Output {
+        Fee(self + rhs.0)
+    }
+}
+
+impl Add<u128> for Fee {
+    type Output = Fee;
+
+    fn add(self, rhs: u128) -> Self::Output {
+        Fee(self.0 + rhs)
+    }
+}
+
+impl Mul<Fee> for Fee {
+    type Output = Fee;
+
+    fn mul(self, rhs: Fee) -> Self::Output {
+        Fee(self.0 * rhs.0)
+    }
+}
+
+impl Mul<Fee> for u128 {
+    type Output = Fee;
+
+    fn mul(self, rhs: Fee) -> Self::Output {
+        Fee(self * rhs.0)
+    }
+}
+
+impl Mul<u128> for Fee {
+    type Output = Fee;
+
+    fn mul(self, rhs: u128) -> Self::Output {
+        Fee(self.0 * rhs)
+    }
+}
+
+impl Div<Fee> for Fee {
+    type Output = Fee;
+
+    fn div(self, rhs: Fee) -> Self::Output {
+        Fee(self.0 / rhs.0)
+    }
+}
+
+impl Div<Fee> for u128 {
+    type Output = Fee;
+
+    fn div(self, rhs: Fee) -> Self::Output {
+        Fee(self / rhs.0)
+    }
+}
```

Comment:
This implementation also should not exist. The output would be an "inverse fee" (or similar nonsense concept). See my rant about `Fee * Fee` for more details about why getting the semantics right is important.